### ISSUE TITLE
Improve parts API with a Part class

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#885]
instead of using a bunch of `std::vector<bool>`s to indicate state of different parts owned by IBFEMethod we should just implement `class Part` and make each part responsible for storing this information.

**Comment by drwells** [comment]
I took this approach when writing the deal.II bindings:
 https://github.com/drwells/fiddle/blob/master/include/fiddle/mechanics/part.h

But, in retrospect - this involves completely rewriting FEMechanicsBase for not a lot of gain. It would improve the code organization but ultimately it would require changing the user-facing API for no great benefit.

**Comment by boyceg** [comment]
Can we support both APIs for a release or two?

**Comment by drwells** [comment]
By 'both APIs', do you mean the new deal.II one and the current libMesh one? If so, absolutely - we should maintain the libMesh API indefinitely, just not massively rewrite it.

**Comment by boyceg** [comment]
No, I mean add `Part` to the libMesh-based FE API. I would like the deal.II and libMesh-based FE stuff to be as similar as possible.

**Comment by drwells** [comment]
Ah. Maybe? That's going to be harder to do since libMesh has such a fundamentally different set of features. Lets reopen this for now.

**Comment by boyceg** [comment]
In terms of basic model specification, doesn't the deal.II implementation basically support a subset of the features in the libMesh-based implementation (since we have found that some of the stuff offered in the current `IBFEMethod` is not really useful)?

**Comment by drwells** [comment]
Yes. Several things are still missing (like body forces and pressures) but we now enforce that the same FE space is used for displacement, velocity, and force. However, the deal.II implementation goes further than that and basically implements the majority of open libMesh issues here, e.g.,:
- a `Part` class (this issue)
- custom quadrature support (almost done) - #516 
- avoid setting up multiple linear solvers - #1159 
- PK1 performance issues - #996 (though I am mostly done with the libMesh version of that)
- static system names - #1212 these don't exist in deal.II

and some other things (better support for multipart models, automatic secondary hierarchy, etc).

A lot of this just doesn't have analogies in libMesh code. For example, we now solve linear systems like
```cpp
    for (unsigned int part_n = 0; part_n < parts.size(); ++part_n)
      {
        SolverControl control(1000, 1e-14 * rhs_vecs[part_n].l2_norm());
        SolverCG<LinearAlgebra::distributed::Vector<double>> cg(control);
        LinearAlgebra::distributed::Vector<double>           velocity(
          parts[part_n].get_partitioner());
        // TODO - implement initial guess stuff here
        cg.solve(parts[part_n].get_mass_operator(),
                 velocity,
                 rhs_vecs[part_n],
                 parts[part_n].get_mass_preconditioner());
      }
```
because a part can only have one mass operator (which is matrix-free) and one preconditioner. Here, a `Partitioner` is, roughly, a `PetscSF` which has no direct libMesh analog either.

In most of these cases, it wouldn't be that difficult to make the deal.II bindings look more like libMesh (with the exception of static system names) but going in the opposite direction would be difficult.

**Comment by boyceg** [comment]
I am mostly thinking in terms of the APIs for high-level model specification (which are primarily stress and force functions) rather than method/solver specification. I think it is fine (and probably good) for solver stuff and internal organization to be different.

An alternative would be to add in a new API for the libMesh-based implementation that looks like the deal.II implementation, and to deprecate the old API. I am 100% OK with this approach as well.

Basically, to the extent possible, I want to minimize the differences in setting up a model between the two implementations. There are at least two reasons that this seems worthwhile:

1. It makes it easier for users to switch between implementations.
2. It reduces the number of "concepts" that need to be introduced to describe how to set up a model (e.g. in the documentation that I will write some day).

**Comment by drwells** [comment]
This happened, but in Cardinal, not IBAMR.
